process/block/pendingMb/pendingMiniBlocks.go

```diff
@@ -58,114 +48,98 @@ func (p *pendingMiniBlocksHeaders) getAllCrossShardMiniBlocks(metaBlock *block.M
+			crossShardMiniBlocks[string(mbHeader.Hash)] = mbHeader.ReceiverShardID
 		}
 	}
 
 	return crossShardMiniBlocks
 }
 
-// AddProcessedHeader will add all miniblocks headers in a map
-func (p *pendingMiniBlocksHeaders) AddProcessedHeader(headerHandler data.HeaderHandler) error {
+// AddProcessedHeader will add in pending list all miniblocks hashes from a given metablock
+func (p *pendingMiniBlocks) AddProcessedHeader(headerHandler data.HeaderHandler) error {
+	if check.IfNil(headerHandler) {
+		return process.ErrNilHeaderHandler
+	}
+
+	log.Trace("AddProcessedHeader",
+		"shard", headerHandler.GetShardID(),
+		"epoch", headerHandler.GetEpoch(),
+		"round", headerHandler.GetRound(),
+		"nonce", headerHandler.GetNonce())
+
+	return p.processHeader(headerHandler)
+}
+
+// RevertHeader will remove from pending list all miniblocks hashes from a given metablock
+func (p *pendingMiniBlocks) RevertHeader(headerHandler data.HeaderHandler) error {
 	if check.IfNil(headerHandler) {
-		return epochStart.ErrNilHeaderHandler
+		return process.ErrNilHeaderHandler
 	}
 
+	log.Trace("RevertHeader",
+		"shard", headerHandler.GetShardID(),
+		"epoch", headerHandler.GetEpoch(),
+		"round", headerHandler.GetRound(),
+		"nonce", headerHandler.GetNonce())
+
+	return p.processHeader(headerHandler)
+}
+
+func (p *pendingMiniBlocks) processHeader(headerHandler data.HeaderHandler) error {
 	metaBlock, ok := headerHandler.(*block.MetaBlock)
 	if !ok {
-		return epochStart.ErrWrongTypeAssertion
+		return process.ErrWrongTypeAssertion
 	}
 
-	crossShardMiniBlocks := p.getAllCrossShardMiniBlocks(metaBlock)
+	crossShardMiniBlocksHashes := p.getAllCrossShardMiniBlocksHashes(metaBlock)
 
-	p.mutPending.Lock()
-	for mbHash, mbHeader := range crossShardMiniBlocks {
-		if _, ok = p.mapMiniBlockHeaders[mbHash]; !ok {
-			p.mapMiniBlockHeaders[mbHash] = mbHeader
-			p.mapShardNumMiniBlocks[mbHeader.ReceiverShardID]++
-			continue
-		}
+	p.mutPendingMbShard.Lock()
+	defer p.mutPendingMbShard.Unlock()
 
-		delete(p.mapMiniBlockHeaders, mbHash)
-		if p.mapShardNumMiniBlocks[mbHeader.ReceiverShardID] == 0 {
-			log.Error("AddProcessedHeader: trying to remove an unpending miniblock",
-				"epoch", metaBlock.Epoch,
-				"round", metaBlock.Round,
-				"nonce", metaBlock.Nonce,
-				"mb hash", mbHash,
-				"mb sender shard", mbHeader.SenderShardID,
-				"mb receiver shard", mbHeader.ReceiverShardID)
+	for mbHash, shardID := range crossShardMiniBlocksHashes {
+		if _, ok = p.mapPendingMbShard[mbHash]; !ok {
+			p.mapPendingMbShard[mbHash] = shardID
 			continue
 		}
-		p.mapShardNumMiniBlocks[mbHeader.ReceiverShardID]--
-	}
 
-	for shardID, numPendingMiniBlocks := range p.mapShardNumMiniBlocks {
-		log.Trace("pending miniblocks", "shard", shardID, "num", numPendingMiniBlocks)
+		delete(p.mapPendingMbShard, mbHash)
 	}
 
-	p.mutPending.Unlock()
+	for mbHash, shardID := range p.mapPendingMbShard {
+		log.Debug("pending miniblocks", "shard", shardID, "hash", mbHash)
```

Comment:
Trace

Reply:
Yes, I will put it back after testing will be finished